### ISSUE TITLE
openssh-server closes immediately

### ISSUE THREAD

**Issue by Steve-Newcomb** [openwrt/openwrt#10588]
I have installed both dropbear and openssh-server/openssh-client, etc.  `dropbear` is listening at port 20022, and it works.  `sshd` is listening at 22 and a couple of other ports as well.  I do not understand why both client and server fail to communicate with each other, each blaming the other, but only AFTER the public key is accepted.  The agreement on the public key suggests (to me, anyway) that they are communicating OK.

Maybe the firewall isn't getting the message about the port that will be used afterward?  After I do an `/etc/init.d/firewall stop`, the problem persists, so I think not (which may be because I'm making a wrong assumption about the firewall).

The same configuration still works under a build of 21.02.1 that I made some time ago, but it doesn't work now.  I don't know where to look or what to try next.  The test router's hostname is `rpc151`, which may help you decipher the attached (verbose) files, `sshd_server_output.txt`, `ssh_client_output.txt`, `etc-ssh-ssh_config.txt` and `etc-ssh-sshd_config.txt`.  If there's a clue in any of it, I'm certainly not seeing it, but I'm not an openssh expert.

(Alas, the file attachment feature of this github  website keeps saying "Something went really wrong...", so I can only insert the files wholesale right here:)

--------------------------------------------------------
sshd_server_output.txt

root@rpc151:~# /usr/sbin/sshd -d
debug1: sshd version OpenSSH_9.0, OpenSSL 1.1.1q  5 Jul 2022
debug1: private host key #0: ssh-ed25519 SHA256:3pVU9cKYJFNgl+JxXEt6moFeudEuzH4fy7gbzhCT4SQ
debug1: private host key #1: ssh-rsa SHA256:Xf7HrvwwoH5CXObPT+/IKh8JSlGgUtIVWlWq3YCc+fg
debug1: rexec_argv[0]='/usr/sbin/sshd'
debug1: rexec_argv[1]='-d'
debug1: Set /proc/self/oom_score_adj from 0 to -1000
debug1: Bind to port 2751 on ::.
Server listening on :: port 2751.
debug1: Bind to port 2351 on ::.
Server listening on :: port 2351.
debug1: Bind to port 22 on ::.
Server listening on :: port 22.
debug1: Bind to port 2751 on 0.0.0.0.
Server listening on 0.0.0.0 port 2751.
debug1: Bind to port 2351 on 0.0.0.0.
Server listening on 0.0.0.0 port 2351.
debug1: Bind to port 22 on 0.0.0.0.
Server listening on 0.0.0.0 port 22.
debug1: Server will not fork when running in debugging mode.
debug1: rexec start in 9 out 9 newsock 9 pipe -1 sock 12
debug1: sshd version OpenSSH_9.0, OpenSSL 1.1.1q  5 Jul 2022
debug1: private host key #0: ssh-ed25519 SHA256:3pVU9cKYJFNgl+JxXEt6moFeudEuzH4fy7gbzhCT4SQ
debug1: private host key #1: ssh-rsa SHA256:Xf7HrvwwoH5CXObPT+/IKh8JSlGgUtIVWlWq3YCc+fg
debug1: inetd sockets after dupping: 3, 3
Connection from ::1 port 56074 on ::1 port 22 rdomain ""
debug1: Local version string SSH-2.0-OpenSSH_9.0
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0
debug1: compat_banner: match: OpenSSH_9.0 pat OpenSSH* compat 0x04000000
debug1: permanently_set_uid: 22/22 [preauth]
debug1: list_hostkey_types: ssh-ed25519,rsa-sha2-512,rsa-sha2-256 [preauth]
debug1: SSH2_MSG_KEXINIT sent [preauth]
debug1: SSH2_MSG_KEXINIT received [preauth]
debug1: kex: algorithm: curve25519-sha256 [preauth]
debug1: kex: host key algorithm: ssh-ed25519 [preauth]
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]
debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]
debug1: SSH2_MSG_KEX_ECDH_INIT received [preauth]
debug1: rekey out after 134217728 blocks [preauth]
debug1: SSH2_MSG_NEWKEYS sent [preauth]
debug1: Sending SSH2_MSG_EXT_INFO [preauth]
debug1: expecting SSH2_MSG_NEWKEYS [preauth]
debug1: SSH2_MSG_NEWKEYS received [preauth]
debug1: rekey in after 134217728 blocks [preauth]
debug1: KEX done [preauth]
debug1: userauth-request for user root service ssh-connection method none [preauth]
debug1: attempt 0 failures 0 [preauth]
debug1: user root matched 'User root' at line 85
debug1: userauth-request for user root service ssh-connection method publickey [preauth]
debug1: attempt 1 failures 0 [preauth]
debug1: userauth_pubkey: publickey test pkalg rsa-sha2-512 pkblob RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA [preauth]
debug1: temporarily_use_uid: 0/0 (e=0/0)
debug1: trying public key file /root/.ssh/authorized_keys
debug1: fd 4 clearing O_NONBLOCK
debug1: /root/.ssh/authorized_keys:13: matching key found: RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
debug1: /root/.ssh/authorized_keys:13: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
Accepted key RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA found at /root/.ssh/authorized_keys:13
debug1: restore_uid: 0/0
Postponed publickey for root from ::1 port 56074 ssh2 [preauth]
debug1: userauth-request for user root service ssh-connection method publickey-hostbound-v00@openssh.com [preauth]
debug1: attempt 2 failures 0 [preauth]
debug1: temporarily_use_uid: 0/0 (e=0/0)
debug1: trying public key file /root/.ssh/authorized_keys
debug1: fd 4 clearing O_NONBLOCK
debug1: /root/.ssh/authorized_keys:13: matching key found: RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
debug1: /root/.ssh/authorized_keys:13: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
Accepted key RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA found at /root/.ssh/authorized_keys:13
debug1: restore_uid: 0/0
debug1: auth_activate_options: setting new authentication options
Accepted publickey for root from ::1 port 56074 ssh2: RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
debug1: monitor_child_preauth: user root authenticated by privileged process
debug1: auth_activate_options: setting new authentication options [preauth]
debug1: monitor_read_log: child log fd closed
debug1: rekey in after 134217728 blocks
debug1: rekey out after 134217728 blocks
debug1: ssh_packet_set_postauth: called
debug1: active: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
debug1: Entering interactive session for SSH2.
debug1: server_init_dispatch
debug1: server_input_channel_open: ctype session rchan 0 win 1048576 max 16384
debug1: input_session_request
debug1: channel 0: new [server-session]
debug1: session_new: session 0
debug1: session_open: channel 0
debug1: session_open: session 0: link with channel 0
debug1: server_input_channel_open: confirm session
debug1: server_input_global_request: rtype no-more-sessions@openssh.com want_reply 0
Timeout, client not responding from user root ::1 port 56074
debug1: do_cleanup
root@rpc151:~# 

--------------------------------------------------------
ssh_client_output.txt
root@rpc151:~# ssh -v localhost
OpenSSH_9.0p1, OpenSSL 1.1.1q  5 Jul 2022
debug1: Reading configuration data /root/.ssh/config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug1: Connecting to localhost [::1] port 22.
debug1: fd 3 clearing O_NONBLOCK
debug1: Connection established.
debug1: identity file /root/.ssh/id_rsa type 0
debug1: identity file /root/.ssh/id_rsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa type -1
debug1: identity file /root/.ssh/id_ecdsa-cert type -1
debug1: identity file /root/.ssh/id_ecdsa_sk type -1
debug1: identity file /root/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /root/.ssh/id_ed25519 type -1
debug1: identity file /root/.ssh/id_ed25519-cert type -1
debug1: identity file /root/.ssh/id_ed25519_sk type -1
debug1: identity file /root/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /root/.ssh/id_xmss type -1
debug1: identity file /root/.ssh/id_xmss-cert type -1
debug1: identity file /root/.ssh/id_dsa type -1
debug1: identity file /root/.ssh/id_dsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_9.0
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0
debug1: compat_banner: match: OpenSSH_9.0 pat OpenSSH* compat 0x04000000
debug1: Authenticating to localhost:22 as 'root'
debug1: load_hostkeys: fopen /root/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ssh-ed25519
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: SSH2_MSG_KEX_ECDH_REPLY received
debug1: Server host key: ssh-ed25519 SHA256:3pVU9cKYJFNgl+JxXEt6moFeudEuzH4fy7gbzhCT4SQ
debug1: load_hostkeys: fopen /root/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: Host 'localhost' is known and matches the ED25519 host key.
debug1: Found key in /root/.ssh/known_hosts:1
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /root/.ssh/id_rsa RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
debug1: Will attempt key: /root/.ssh/id_ecdsa 
debug1: Will attempt key: /root/.ssh/id_ecdsa_sk 
debug1: Will attempt key: /root/.ssh/id_ed25519 
debug1: Will attempt key: /root/.ssh/id_ed25519_sk 
debug1: Will attempt key: /root/.ssh/id_xmss 
debug1: Will attempt key: /root/.ssh/id_dsa 
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,webauthn-sk-ecdsa-sha2-nistp256@openssh.com>
debug1: kex_input_ext_info: publickey-hostbound@openssh.com=<0>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering public key: /root/.ssh/id_rsa RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
debug1: Server accepts key: /root/.ssh/id_rsa RSA SHA256:C3EMnwJej8S6dYBvWrtXP9FrdA2u4C7H96r0MS217lA
Authenticated to localhost ([::1]:22) using "publickey".
debug1: channel 0: new [client-session]
debug1: Requesting no-more-sessions@openssh.com
debug1: Entering interactive session.
debug1: pledge: filesystem
debug1: client_input_global_request: rtype hostkeys-00@openssh.com want_reply 0
debug1: client_input_hostkeys: searching /root/.ssh/known_hosts for localhost / (none)
debug1: client_input_hostkeys: searching /root/.ssh/known_hosts2 for localhost / (none)
debug1: client_input_hostkeys: hostkeys file /root/.ssh/known_hosts2 does not exist
debug1: client_input_hostkeys: no new or deprecated keys from server
debug1: Remote: /root/.ssh/authorized_keys:13: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
debug1: Remote: /root/.ssh/authorized_keys:13: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
debug1: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
debug1: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
debug1: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
debug1: channel 0: free: client-session, nchannels 1
Connection to localhost closed by remote host.
Connection to localhost closed.
Transferred: sent 3248, received 2416 bytes, in 0.0 seconds
Bytes per second: sent 127467.8, received 94816.0
debug1: Exit status -1
root@rpc151:~# 

--------------------------------------------------------
etc-ssh-ssh_config.txt

# Host *
#   ForwardAgent no
#   ForwardX11 no
#   PasswordAuthentication yes
#   HostbasedAuthentication no
#   GSSAPIAuthentication no
#   GSSAPIDelegateCredentials no
#   BatchMode no
#   CheckHostIP yes
#   AddressFamily any
#   ConnectTimeout 0
#   StrictHostKeyChecking ask
#   IdentityFile ~/.ssh/id_rsa
#   IdentityFile ~/.ssh/id_dsa
#   IdentityFile ~/.ssh/id_ecdsa
#   IdentityFile ~/.ssh/id_ed25519
#   Port 22
#   Ciphers aes128-ctr,aes192-ctr,aes256-ctr,aes128-cbc,3des-cbc
#   MACs hmac-md5,hmac-sha1,umac-64@openssh.com
#   EscapeChar ~
#   Tunnel no
#   TunnelDevice any:any
#   PermitLocalCommand no
#   VisualHostKey no
#   ProxyCommand ssh -q -W %h:%p gateway.example.com
#   RekeyLimit 1G 1h
#   UserKnownHostsFile ~/.ssh/known_hosts.d/%k

--------------------------------------------------------
etc-ssh-sshd_config.txt

#AcceptEnv LANG LC_*
AddressFamily any
AllowAgentForwarding yes
AllowTcpForwarding yes
#AuthorizedKeysCommand none
AuthorizedKeysCommandUser nobody
AuthorizedKeysFile	.ssh/authorized_keys
AuthorizedPrincipalsFile none
Banner none
#ChallengeResponseAuthentication no
###Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
ChrootDirectory none
ClientAliveCountMax 3
ClientAliveInterval 30
Compression delayed
#DebianBanner no
#GSSAPIAuthentication no
#GSSAPICleanupCredentials yes
GatewayPorts clientspecified
#HostKey /etc/ssh/ssh_host_ecdsa_key
HostKey /etc/ssh/ssh_host_ed25519_key
HostKey /etc/ssh/ssh_host_rsa_key
###HostKeyAlgorithms sk-ssh-ed25519@openssh.com,ssh-ed25519,ssh-ed25519-cert-v01@openssh.com,rsa-sha2-256,rsa-sha2-512,rsa-sha2-256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com
HostbasedAuthentication no
IgnoreRhosts yes
IgnoreUserKnownHosts no
KbdInteractiveAuthentication yes
#KerberosAuthentication no
#KerberosGetAFSToken no
#KerberosOrLocalPasswd yes
#KerberosTicketCleanup yes
##KexAlgorithms curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group14-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha256
ListenAddress 0.0.0.0
ListenAddress ::
LogLevel DEBUG
LoginGraceTime 30
###MACs hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128-etm@openssh.com
MaxAuthTries 6
MaxSessions 10
MaxStartups 10:30:100
PasswordAuthentication yes
PermitEmptyPasswords no
PermitRootLogin yes
PermitTTY yes
PermitTunnel no
PermitUserEnvironment no
PidFile /var/run/sshd.pid
Port 22
Port 2351
Port 2751
## Port 8551
#PrintLastLog yes
PrintMotd no
#Protocol 2
PubkeyAuthentication yes
#RekeyLimit default none
StrictModes yes
Subsystem	sftp	/usr/lib/sftp-server
SyslogFacility AUTH
TCPKeepAlive yes
UseDNS yes
###
#UsePAM no
UsePAM yes
###
#VersionAddendum none
X11DisplayOffset 10
X11Forwarding no
#X11UseLocalhost yes
Match User keepalive
    AllowTcpForwarding yes
    X11Forwarding yes
    PermitTunnel yes
    GatewayPorts yes
    AllowAgentForwarding yes
    PasswordAuthentication no
    AuthorizedKeysFile	/home/keepalive/.ssh/authorized_keys

Match User meshmaint
    AllowTcpForwarding yes
    X11Forwarding yes
    PermitTunnel yes
    GatewayPorts yes
    AllowAgentForwarding yes
    PasswordAuthentication no
    AuthorizedKeysFile	/home/meshmaint/.ssh/authorized_keys

###Match User root

**Comment by Hurricos** [comment]
`cat /etc/config/openssl`?

**Comment by Steve-Newcomb** [comment]
There is no such file as /etc/config/openssl, nor is there in the earlier (working) build of 21.02.1.  (Interesting idea, though.)

**Comment by Hurricos** [comment]
I was wondering if you'd perhaps been running hardware accelerated OpenSSL.
Martin

**Comment by Steve-Newcomb** [comment]
Fixed.  After a lot of experimentation, it turned out that the problem was in `/etc/ssh/sshd_config`.

The problem was:

`ClientAliveInterval 30`

That specification no longer works in OpenWRT, although it worked a few months ago on a fresh build of 21.02.1.  Nowadays, the ClientAliveInterval specification, if it is uncommented, *must* be `ClientAliveInterval 0`.  I don't know why it stopped working.  If any value other than the default value 0 is used, public keys don't work.  In my configuration, the server falls back to giving a password prompt.  No reason for the public key failure is logged, even with `LogLevel DEBUG`.  

Who'd have guessed *that* spec was the source of the trouble?  In order to figure this out, I was reduced to a tedious line-by-line "does it still work?" process.
